### PR TITLE
feat: improve http network filters

### PR DESCRIPTION
**Pull request by aymericDD**
## What does this PR do?

- [ ] Adds new functionality
- [x] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- **CHAOS-738 CHAOS-739**: Allow user to define a list of methods and paths to apply a network disruption:

> Before

**TC architecture**

image

**network_filters_http.yml**

``` yaml
# Unless explicitly stated otherwise all files in this repository are licensed
# under the Apache License Version 2.0.
# This product includes software developed at Datadog (https://www.datadoghq.com/).
# Copyright 2023 Datadog, Inc.

apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: network-drop
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 1
  network:
    drop: 100 # percentage of outgoing packets to drop
    http: # http filters
      method: ALL # http method to filter on
      path: / # requested path to filter on

```

> After

**TC architecture**

image

**network_filters_http.yml**

``` yaml
# Unless explicitly stated otherwise all files in this repository are licensed
# under the Apache License Version 2.0.
# This product includes software developed at Datadog (https://www.datadoghq.com/).
# Copyright 2023 Datadog, Inc.

apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: network-drop
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 1
  network:
    drop: 100 # percentage of outgoing packets to drop
    http: # http filters
      methods: 
        - GET # http method to filter on
      paths:
        - / # requested path to filter on
```

- **CHAOS-795**: Use the libbpfgo library to update eBPF maps of the http network filters.

## Code Quality Checklist

- [ ] The documentation is up to date.
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [x] I manually tested the following steps:
    - `x`
    - [x] locally
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: docs/network_disruption/prio.md [hunk 1]
```diff
@@ -205,8 +207,10 @@ spec:
     port: 80
     protocol: tcp
     http:
-        method: get <-- eBPF filter
-        path: /test <-- eBPF filter
+        methods:
```
Comment (ptnapoleon) [inline docs/network_disruption/prio.md:210]: I have no objections to choosing this approach, but why didn't we decide to let users specify methods _per_ path? Does it create too many tc filters?

Comment (aymericDD) [reply docs/network_disruption/prio.md:210]: I have tried to implement this kind of feature, but it was really hard (memory/complexity) to pair paths with methods so after 1 week I have abandoned this idea. 😞

Comment (ptnapoleon) [reply docs/network_disruption/prio.md:210]: That's a good enough reason for me.

File: injector/network_disruption.go [hunk 2]
```diff
@@ -309,12 +309,15 @@ func (i *networkDisruptionInjector) Clean() error {
 //   - a first prio qdisc will be created and attached to root
 //     it'll be used to apply the first filter, filtering on packet IP destination, source/destination ports and protocol
 //   - a second prio qdisc will be created and attached to the first one
-//     it'll be used to apply the second eBPF filter, filtering on method and path
+//     it'll be used to apply the second eBPF filter, filtering on method
 //   - a third prio qdisc will be created and attached to the second one
+//     it'll be used to apply the second eBPF filter, filtering on path
+//   - a fourth prio qdisc will be created and attached to the second one
```
Comment (ptnapoleon) [inline injector/network_disruption.go:315]: Isn't it attached to the third one, in the diagram?

File: injector/network_disruption.go [hunk 3]
```diff
@@ -309,12 +309,15 @@ func (i *networkDisruptionInjector) Clean() error {
...
 //     it'll be used to apply the third filter, filtering on packet mark to identify packets coming from the targeted process
 //   - operations will be chained to the third band of the third prio qdisc
 //   - an fw filter will be created to classify packets according to their mark (if any)
-//   - an eBOF filter will be created to classify packets according to their method and path (if any)
+//   - a first eBPF filter will be created to classify packets according to their method (if any)
+//   - a second eBPF filter will be created to classify packets according to their  path (if any)
```
Comment (ptnapoleon) [inline injector/network_disruption.go:320]:
```suggestion
//   - a second eBPF filter will be created to classify packets according to their path (if any)
```

File: api/v1beta1/network_disruption.go [hunk 4]
```diff
@@ -150,23 +172,118 @@ type NetworkDisruptionCloudServiceSpec struct {
...
 }
 
-// Validate validates args for the given http filters.
-func (s *NetworkHTTPFilters) Validate() error {
-	if s.Path != "" {
-		if len(s.Path) > MaxNetworkPathCharacters {
-			return fmt.Errorf("the path specification at the network disruption level is not valid; should not exceed 100 characters")
+func (p HTTPPath) validate() error {
+	if len(p) > MaxNetworkPathCharacters {
+		return fmt.Errorf("the paths specification at the network disruption level is not valid; should not exceed 100 characters")
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:177]:
```suggestion
		return fmt.Errorf("the paths specification at the network disruption level is not valid; should not exceed %d characters", MaxNetworkPathCharacters)
```

File: api/v1beta1/network_disruption.go [hunk 5]
```diff
@@ -150,23 +172,118 @@ type NetworkDisruptionCloudServiceSpec struct {
...
+}
+
+func (h HTTPMethods) isNotEmpty() bool {
+	return len(h) >= 1
+}
+
+// validate validates args for the given http filters.
+func (s *NetworkHTTPFilters) validate() (retErr error) {
+	if s.DeprecatedPath != "" {
+		retErr = multierror.Append(retErr, fmt.Errorf("the Path specification at the HTTP network disruption level is deprecated; use Paths HTTP field instead"))
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:206]:
```suggestion
		retErr = multierror.Append(retErr, fmt.Errorf("the Path specification at the HTTP network disruption level is no longer supported; use Paths HTTP field instead"))
```

it's fine to label these as deprecated internally, but technically deprecated means we are still supporting something "for now" but will remove it later. However, these are already removed, so we should just be clear

File: api/v1beta1/network_disruption.go [hunk 6]
```diff
@@ -150,23 +172,118 @@ type NetworkDisruptionCloudServiceSpec struct {
...
+	}
+
+	if s.DeprecatedMethod != "" {
+		retErr = multierror.Append(retErr, fmt.Errorf("the Method specification at the HTTP network disruption level is deprecated; use Methods HTTP field instead"))
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:210]:
```suggestion
		retErr = multierror.Append(retErr, fmt.Errorf("the Method specification at the HTTP network disruption level is no longer supported; use Methods HTTP field instead"))
```

File: api/v1beta1/network_disruption.go [hunk 7]
```diff
@@ -150,23 +172,118 @@ type NetworkDisruptionCloudServiceSpec struct {
...
+	}
+
+	if len(s.Paths) > MaxNetworkPaths {
+		retErr = multierror.Append(retErr, fmt.Errorf(HTTPPathsFilterErrorPrefix+"the number of paths must not be greater than %d; Number of paths: %d", MaxNetworkPaths, len(s.Paths)))
+	} else if len(s.Paths) > 0 {
+		visitedPaths := make(map[HTTPPath]struct {
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:216]: Could we break this entire block under `else if len(s.Paths) > 0` out into a function w/ unit tests?

File: api/v1beta1/network_disruption.go [hunk 8]
```diff
@@ -150,23 +172,118 @@ type NetworkDisruptionCloudServiceSpec struct {
...
+			}
+
+			visitedMethod.count++
+			visitedMethods[method] = visitedMethod
+		}
+
+		for method, visitedMethod := range visitedMethods {
+			if visitedMethod.count > 1 {
+				retErr = multierror.Append(retErr, fmt.Errorf(HTTPMethodsFilterErrorPrefix+"should not contain duplicated methods. Count: %d; Method: %s", visitedMethod.count, method))
+				delete(visitedMethods, method)
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:281]:
```suggestion
```

I don't think we should mutate this while we're iterating through it. It also shouldn't have an effect on our behavior regardless? We'll only list one error per method

File: injector/network_disruption.go [hunk 9]
```diff
@@ -423,33 +429,50 @@ func (i *networkDisruptionInjector) applyOperations() error {
...
+
+			if err = i.config.TrafficController.ConfigBPFFilter(bpfConfigExecutor, configBPFFilterArgs...); err != nil {
 				return fmt.Errorf("could not update the configuration of the bpf-network-tc-filter filter: %w", err)
 			}
 
-			// parent 3:2 refers to the 2nd band of the 3nd prio qdisc
-			// handle starts from 4 because 1, 2 are used by the 3 prio qdiscs
-			parent = "3:2"
-			handle = uint32(4)
+			// parent 4:2 refers to the 3nd band of the 4nd prio qdisc
```
Comment (ptnapoleon) [inline injector/network_disruption.go:472]:
```suggestion
			// parent 4:2 refers to the 3nd band of the 4th prio qdisc
```

File: api/v1beta1/network_disruption.go (hunk 7 above)
Comment (aymericDD) [reply api/v1beta1/network_disruption.go:216]: A good practice is to unit tests only public functions. Private functions are tested implicitely and usually come from the refactoring step. In this situation the unit tests are dedicated to the public function `Validate` of the `NetworkDisruptionSpec`.  Source https://sd.blackball.lv/library/unit_testing_(2020).pdf page 260 `11.1 Unit testing private methods`

File: injector/network_disruption.go (hunk 2 above)
Comment (aymericDD) [reply injector/network_disruption.go:315]: 🙈 Indeed...

File: api/v1beta1/network_disruption.go (hunk 5 above)
Comment (aymericDD) [reply api/v1beta1/network_disruption.go:206]: Completely agree with you! It's a mistake from my part...

File: api/v1beta1/network_disruption.go (hunk 8 above)
Comment (aymericDD) [reply api/v1beta1/network_disruption.go:281]: That's true! I 👍

File: injector/network_disruption.go [hunk 10]
```diff
@@ -423,33 +429,50 @@ func (i *networkDisruptionInjector) applyOperations() error {
 		}
 
 		if i.spec.HasHTTPFilters() {
-			// create a third prio with only 2 bands to filter traffic with a specific mark
+			// create a third prio with only 2 bands to filter traffic based on http method
```
[the remark this answers is not in the document]
Comment (aymericDD) [reply injector/network_disruption.go:432]: Indeed, comments are not correct. I have updated the code to be aligned with the reality 👍

File: injector/network_disruption.go [hunk 11]
```diff
@@ -421,40 +427,63 @@ func (i *networkDisruptionInjector) applyOperations() error {
...
+				return fmt.Errorf("can't create the fw filter: %w", err)
+			}
+
+			// parent 4:2 refers to the 3nd band of the 4th prio qdisc
+			// handle starts from 5 because 1, 2 and 3 are used by the 4 prio qdiscs
+			parent = "4:2"
+			handle = uint32(5)
 		} else {
+			// create second prio with only 2 bands to filter traffic with a specific mark
+			if err := i.config.TrafficController.AddPrio(interfaces, "1:4", "2:", 2, [16]uint32{}); err != nil {
```
Comment (ptnapoleon) [inline injector/network_disruption.go:483]: Why do we need this new Prio for non-http filters?

Comment (ptnapoleon) [reply injector/network_disruption.go:483]: I see this entered in `126e17192ffc5293ddd17823b52e9de0f4894ca3`, seems unexpected?

Comment (aymericDD) [reply injector/network_disruption.go:483]: Before this commit this line was applied for both scenarios: 
**Without http filters**: this qdisc is used to check if the mark is applied
**With http filters**: this qdisc is used to check the http method.
So I duplicated this line and add different comments depending of the scenario.

**Review by ptnapoleon — APPROVED** [review]

**Pull request merged**